File: javascript/secret-handshake/secret-handshake.js
'use strict';

var binary = require('../binary/binary');

var OPERATIONS = ['wink', 'double blink', 'close your eyes', 'jump'];

function SecretHandshake(number) {
    if (typeof(number) !== 'number') {
        throw new Error("Handshake must be a number");
    }
    this.number = number;
}

SecretHandshake.prototype.commands = function () {
    var self = this;
    // apply a bit-mask, based on the index of the operation
    var commands = OPERATIONS.filter(function (op, index) {
        return self.number & (1 << index);
    });
    // finally, reverse it if it matches the next binary step
    if (self.number & (1 << (OPERATIONS.length))) {
        commands.reverse();
    }
    return commands;
};

module.exports = SecretHandshake;

File: elixir/secret-handshake/lib/secret_handshake.ex
defmodule SecretHandshake do
  @commands [
    {8, "jump"},
    {4, "close your eyes"},
    {2, "double blink"},
    {1, "wink"},
    {16, :reverse}
  ]
  @doc """
  Determine the actions of a secret handshake based on the binary
  representation of the given `code`.

  If the following bits are set, include the corresponding action in your list
  of commands, in order from lowest to highest.

  1 = wink
  10 = double blink
  100 = close your eyes
  1000 = jump

  10000 = Reverse the order of the operations in the secret handshake
  """
  @spec commands(code :: integer) :: list(String.t())
  def commands(code) do
    Enum.reduce(@commands, [], fn {bit, command}, acc ->
      if Bitwise.band(code, bit) > 0 do
        if command == :reverse do
          Enum.reverse(acc)
        else
          [command | acc]
        end
      else
        acc
      end
    end)
  end
end
